components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +106,368 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
     }
-    self.buttons = [mutableButtons copy];
   }
 }
 
-- (NSUInteger)numberOfButtons {
-  return self.buttons.count;
-}
+// TODO: KVO trailingButton.hidden and call `setNeedsUpdateConstraints:`.
 
 #pragma mark - UIView overrides
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size
-                                                     textLabel:self.textLabel
-                                                imageContainer:self.iconImageViewContainer
-                                                       buttons:self.buttons];
-  return [self.layout frameSize];
+  MDCBannerViewLayoutMode layoutMode = [self layoutModeForSizeToFit:size];
+  CGFloat frameHeight = 0.0f;
+  switch (layoutMode) {
+    case MDCBannerViewLayoutModeSingleLine: {
+      frameHeight += kTopPaddingSmall + kBottomPadding;
+      CGSize textLabelSize = [self.textLabel sizeThatFits:CGSizeZero];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGFloat maximumHeight = MAX(textLabelSize.height, leadingButtonSize.height);
+      if (self.hasImage) {
+        maximumHeight = MAX(kImageViewSideLength, maximumHeight);
+      }
+      frameHeight += maximumHeight;
+      break;
+    }
+    case MDCBannerViewLayoutModeMultiLineAlignedButton: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGSize trailingButtonSize = [self.trailingButton sizeThatFits:CGSizeZero];
+      frameHeight += MAX(leadingButtonSize.height, trailingButtonSize.height);
+      break;
+    }
+    case MDCBannerViewLayoutModeMultiLineStackedButton: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGSize trailingButtonSize = [self.trailingButton sizeThatFits:CGSizeZero];
+      frameHeight +=
+          leadingButtonSize.height + trailingButtonSize.height + kButtonVerticalIntervalSpace;
+      break;
+    }
+  }
+  return CGSizeMake(size.width, frameHeight);
 }
 
-- (void)layoutSubviews {
-  [super layoutSubviews];
+- (void)updateConstraints {
+  MDCBannerViewLayoutMode layoutMode = [self layoutModeForSizeToFit:self.bounds.size];
+  [self updateConstraintsWithLayoutMode:layoutMode];
 
-  [self applyMDCBannerViewLayout:self.layout];
+  [super updateConstraints];
 }
 
 #pragma mark - Layout methods
 
-- (void)applyMDCBannerViewLayout:(MDCBannerViewLayout *)layout {
-  if (![self isMDCBannerViewLayoutApplicable:layout]) {
-    return;
+- (void)updateConstraintsWithLayoutMode:(MDCBannerViewLayoutMode)layoutMode {
+  [self deactivateAllConstraints];
+  // Set Elements
+  if (@available(iOS 9.0, *)) {
```

Comment:
This method (and any others like it where all of the logic is conditional on the platform) can also make use of NS_AVAILABLE_IOS to remove a layer of logical indentation.

Reply:
+1 to marking available at compile time. Unfortunately, our catalogs/examples only seem to work with dynamic checks because we never explicitly depend on these classes when we compile/run.

Reply:
Addressed in 2bfba78355fdac2c698ea362166ec618d1a6aa5f